src/mongoose_rabbit_worker.erl

```diff
@@ -168,11 +182,91 @@ make_chat_msg(From, To, UserMsg) ->
 is_user_online(online) -> true;
 is_user_online(offline) -> false.
 
--spec close_rabbit_connection(Connection :: pid(), Channel :: pid()) ->
+-spec publish_message_and_wait_for_confirm(Channel :: pid(),
+                                           Exchange :: binary(),
+                                           RoutingKey :: binary(),
+                                           Message :: binary()) -> term().
+publish_message_and_wait_for_confirm(Channel, Exchange, RoutingKey, Message) ->
+    amqp_channel:call(Channel, #'basic.publish'{exchange = Exchange,
+                                                routing_key = RoutingKey},
+                      #amqp_msg{payload = Message}),
+    amqp_channel:wait_for_confirms(Channel).
+
+-spec establish_rabbit_connection(Opts :: #amqp_params_network{},
+                                  Host :: jid:server()) -> ok.
+establish_rabbit_connection(Opts, Host) ->
+    case amqp_connection:start(Opts) of
+        {ok, Connection} ->
+            update_success_rabbit_connections_metrics(Host),
+            {ok, Channel} = amqp_connection:open_channel(Connection),
+            ?DEBUG("event=rabbit_connection_established", []),
+            {Connection, Channel};
+        {error, Error} ->
+            update_failed_rabbit_connections_metrics(Host),
+            ?ERROR_MSG("event=rabbit_connection_failed reason=~1000p", [Error]),
+            exit("connection to a Rabbit server failed")
+    end.
+
+-spec enable_confirms(Channel :: pid()) -> ok.
+enable_confirms(Channel) ->
+    #'confirm.select_ok'{} = amqp_channel:call(Channel, #'confirm.select'{}),
+    ok.
+
+-spec close_rabbit_connection(Connection :: pid(), Channel :: pid(),
+                              Host :: jid:server()) ->
     ok | term().
-close_rabbit_connection(Connection, Channel) ->
+close_rabbit_connection(Connection, Channel, Host) ->
+    update_closed_rabbit_connections_metrics(Host),
     try amqp_channel:close(Channel)
     catch
         _Error:_Reason -> already_closed
     end,
     amqp_connection:close(Connection).
+
+-spec update_messages_published_metrics(Host :: jid:server(),
+                                        PublishTime :: non_neg_integer(),
+                                        Message :: binary()) -> any().
+update_messages_published_metrics(Host, PublishTime, Message) ->
```

Comment:
Maybe make `mod_event_pusher_rabbit` provide an API for bumping the metrics?

Anyway, it a [circular dependency](https://en.wikipedia.org/wiki/Circular_dependency) which is [better to be avoided](http://learnyousomeerlang.com/modules).

Why don't you just put appropriate metrics in the worker module? You can always export a function that lists the required metrics and the `mod_event_push_rabbit` would call it to crate them in its `init_metrics/0`.

Reply:
@mentels please see the latest commit.